geps/gep-3155/index.md

```diff
@@ -0,0 +1,264 @@
+- given OOTB software uses optional certificates for end-user authentication, if the
+  certificate is missing, it falls back to a different application-level authentication
+  method - in such case not providing any certificate is the only way to operate
+
+**1. Add a new `BackendTLS` field at the top level of Gateways**
+
+```go
+type GatewaySpec struct {
+  // BackendTLS configures TLS settings for when this Gateway is connecting to
+  // backends with TLS.
+  BackendTLS GatewayBackendTLS `json:"backendTLS,omitempty"'
+}
+
+type GatewayBackendTLS struct {
+  // ClientCertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // ClientCertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // This setting can be overriden on the service level by use of BackendTLSPolicy.
+  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
+}
+```
+
+**2. Add a new `ClientCertificateMode` and `ClientCertificateRef` fields at the top
+level of BackendTLSPolicy**
+
+```go
+// +kubebuilder:validation:Enum=Default;Override;Disable
+type ClientCertificateInheritanceMode string
+
+const (
+  // Default determines the gateway certificate will be used. If no certificate is
+  // configured on the gateway level, no client certificate will be used.
+  //
+  // Support: Core
+  DefaultClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Default"
+
+  // Override determines the certificate specified in ClientCertificateRef field will be used.
+  //
+  // Support: Implementation-specific
+  OverrideClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Override"
+
+  // Disable determines the certificate will be not provided to the backend, even if it is
+  // configured at the gateway level and requested by the backend.
+  //
+  // Support: Implementation-specific
+  DisableClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Disable"
+)
+
+type BackendTLSPolicySpec {
+  // ClientCertificateMode determines the inheritance mode for client certifacate.
+  //
+  // If ClientCertificateMode is unset, the value assumed is Default.
+  ClientCertificateMode ClientCertificateInheritanceMode `json:"clientCertificateMode,omitempty"`
+
+  // ClientCertificateRef is a reference to an object that contains a Client
+  // Certificate.
+  //
+  // This value takes effect and is required only if the ClientCertificateMode is set to 
+  // Override, otherwise it is ignored.
+  //
+  // References to a resource in different namespace are invalid UNLESS there
```

Comment:
Does this mean I can have:
* Gateway in ns1, selecting route
* Route in ns2, targeting svc in ns3
* BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4
* Cert in ns4 with refgrant allowing backendTLS policy

So ultimately I have a chain of 4 namespaces allowing the Gateway is ns1 to read a secret in ns4?

Reply:
That seems valid from what I can see, and is a good example of how quickly complexity can build  out of small changes.

Reply:
Yes, this seems to be a valid scenario, do you perceive this as a problem?

Reply:
Maybe? The principles as I understand it is there should be a strong two way handshake between each cross namespace hop.

So the question we have is did the `ns4` owner authorization `ns1` to read its Secrets.

Working backwards:
* ns4 granted ns3 Secret read access explicitly with RefGrant. :heavy_check_mark: 
* ns3 explicitly granted ns2 access to **send to services**. :stop_sign: 
* ns1 explicitly granted ns2 access to bind routes :heavy_check_mark: 

I feel there is a violation here. There is no strong agreement from ns3 that ns2 can access its Secrets -- only that it can access its services.

Reply:
I agree with this concern - this calls at least for a clear warning when:
- BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4

ns3 owner should be aware that it is exposing the cert to all gateways that are allowed to **send to this service** will be able to access the certificate

Reply:
I don't really think that is sufficient. Today, there may be many many users giving Service access. This is an **extremely low** privilege to grant, so users will hand it out freely.

Now we are saying that in a new release, that the persona is not even aware of, suddenly granting Service (low priv) gives access to Secret (high priv)

Reply:
How about requiring two grants, i.e. ?
- ns4 to ns3 secret (using secret to connect to the backend you control)
- ns1 to ns3 secret (to consume the secret)

Reply:
This is a good point. A few options I can think of:

**1. References to Secrets that are used by the Gateway are relative to that Gateway**
With this option, we would consider the reference to this Secret as coming from the Gateway when it comes to cross-namespace logic. 

That means that a reference from BackendTLSPolicy in ns3 to a cert in ns1 would be considered _local_ and not need a ReferenceGrant, because the _Gateway_ is in ns1. On the other hand, a reference from BackendTLSPolicy in ns3 to a Secret in ns3 would require a ReferenceGrant that allowed a reference from the Gateway in ns1 to the secret in ns3.

**2. References are relative to BackendTLSPolicy, BackendTLSPolicy should be considered a privileged resource**

I've increasingly been making the case that Service owners aren't necessarily the correct owners for BackendTLSPolicy, particularly in larger organizations. If we consider BackendTLSPolicy a privileged resource, I think it's reasonable to say that valid references from this resource inherently grant access to the Gateway(s)  implementing that resource.

**3. Namespace-local for now**
I don't think we have clear evidence that we _need_ cross-namespace references at this level. Given the complexity involved, it may be safer to explicitly prohibit that as a starting point. We could instead require the Gateway, BackendTLSPolicy, and cert to all be in the same namespace when certs are specified at this level. That does admittedly feel rather onerous, but it's easier to add cross-namespace support in the future than to retroactively fix things if we leave things to open-ended to start with.


With all that said, I'm leaning towards the third option here. Open to other perspectives.

Reply:
I like the idea of requiring BackendTLSPolicy to be in the very same namespace as Gateway, but I am afraid this ship has sailed...  as per: https://gateway-api.sigs.k8s.io/api-types/backendtlspolicy/
"The BackendTLSPolicy and the Service must reside in the same namespace in order to prevent the complications involved with sharing trust across namespace boundaries."

Reply:
I believe we should park this discussion until we reach conclusion #3226, I am moving this overrides part to the future work.